AutoHostRewrite

NishantSinghhhhh opened the pull request:
### Description

Add the ability to automatically rewrite the Host header to the upstream DNS name in the `TrafficPolicy`. This exposes Envoy's built-in `auto_host_rewrite` field and simplifies proxying to SaaS services (e.g., S3, APIs, CDNs) that require a matching `Host` header.

Fixes #11208

---

### Change Type

/kind feature

---

### Changelog

Support for automatic Host header rewriting has been added to the `TrafficPolicy` API via the `autoHostRewrite` field. This enables easier proxying to external services that rely on strict DNS-based Host matching.

---

### Additional Notes

- Only valid for `HTTPRoute` targets.  
- Mutually exclusive with `hostRewrite`, `hostRewriteHeader`, etc.  
- Automatically sets `RouteAction.auto_host_rewrite` in the generated Envoy configuration.

---
<pre> ```release-note Support for automatic Host header rewriting has been added to the `TrafficPolicy` API via the `autoHostRewrite` field. This enables easier proxying to external services that rely on strict DNS-based Host matching. ``` </pre>

NishantSinghhhhh commented:
**Key hand-edited files**

* **API surface** – `api/v1alpha1/traffic_policy_types.go`
  • added `AutoHostRewrite *bool \`json:"autoHostRewrite,omitempty"\``inside`Forwarding\`.

* **Apply-config builder** – `api/applyconfiguration/api/v1alpha1/trafficpolicyspec.go`
  • new fluent setter `WithAutoHostRewrite()`.

* **Apply-config registration** – `api/applyconfiguration/internal/internal.go`
  • registered the new field so generic builders pick it up.

* **Apply-config struct** – `api/applyconfiguration/api/v1alpha1/trafficpolicy.go`
  • struct now carries the `AutoHostRewrite` pointer.

* **CRD / Helm template** – `install/helm/kgateway-crds/templates/gateway.kgateway.dev_trafficpolicies.yaml`
  • added `autoHostRewrite` boolean schema + `oneOf` exclusivity rule.

* **Translator plugin** – `internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go`
  • sets `routeAction.AutoHostRewrite = wrapperspb.Bool(true)` when enabled.

**Files regenerated by `make generated-code`**

* `api/v1alpha1/zz_generated.deepcopy.go`
* multiple `api/applyconfiguration/...` helpers
* `pkg/generated/openapi/zz_generated.openapi.go`
* client-set / lister / informer code

These were touched automatically when the generators re-ran; no manual review needed.

yuval-k commented:
HI @NishantSinghhhhh i enabled workflows to run; i suspect "static code analysis" will fail, if that happens, you should run `make generated-code` to fix go formatting

yuval-k commented on internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:
```diff
@@ -612,6 +614,39 @@ func (p *trafficPolicyPluginGwPass) ApplyForRoute(ctx context.Context, pCtx *ir.
 		p.setTransformationInChain = true
 	}
 
+	if policy.spec.localRateLimit != nil {
+		pCtx.TypedFilterConfig.AddTypedConfig(localRateLimitFilterNamePrefix, policy.spec.localRateLimit)
+
+		// Add a filter to the chain. When having a rate limit for a route we need to also have a
+		// globally disabled rate limit filter in the chain otherwise it will be ignored.
+		// If there is also rate limit for the listener, it will not override this one.
+		if p.localRateLimitInChain == nil {
+			p.localRateLimitInChain = &localratelimitv3.LocalRateLimit{
+				StatPrefix: localRateLimitStatPrefix,
+			}
+		}
+	}
+
+	if policy.spec.rateLimit != nil && policy.spec.rateLimit.rateLimitActions != nil {
+		// Apply global rate limit actions to the route
+		route := outputRoute.GetRoute()
+		if route != nil {
+			route.RateLimits = append(route.GetRateLimits(), policy.spec.rateLimit.rateLimitActions...)
+			// Also call handleRateLimit here to ensure proper setup in rateLimitPerProvider
+			p.handleRateLimit(&pCtx.TypedFilterConfig, policy.spec.rateLimit)
+		}
+	}
```
we recently move this code to a different function. did you intentionally move it back here? or is this the result of a bad merge?

NishantSinghhhhh replied:
Sir, I have updated the repository with the latest changes from upstream. The issue is no longer present, and the if condition for autoHostRewrite is now correctly included within the ApplyForRoute function.

yuval-k commented on internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:
```diff
@@ -1098,6 +1133,15 @@ func (b *TrafficPolicyBuilder) Translate(
+            outSpec.autoHostRewrite = wrapperspb.Bool(true)
+        }
+
+		for _, err := range errors {
+			contextutils.LoggerFrom(ctx).Error(policyCR.GetNamespace(), policyCR.GetName(), err)
+		}
+		policyIr.spec = outSpec
```
i see these lines appearing in the end of the function.. delete this section?

yuval-k commented on internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:
```diff
@@ -129,6 +129,8 @@ type trafficPolicySpecIr struct {
 	localRateLimit             *localratelimitv3.LocalRateLimit
 	rateLimit                  *RateLimitIR
+	errors                     []error
+	autoHostRewrite 		   *wrapperspb.BoolValue
```
please also add a check in the `Equals(in any) bool` function

NishantSinghhhhh replied:
![image](https://github.com/user-attachments/assets/3707ceea-7c9a-4fc4-a2ba-3c4a1f3c4962)

added in Equals function

yuval-k reviewed (commented):
Thank you for the contribution! left some comments

NishantSinghhhhh commented:
@yuval-k sir, can you please review this PR and tell me if any changes are required, I have added tests in the codebase for auto_host_rewrite

NishantSinghhhhh commented:
@yuval-k sir  @lgadban Sir,, can you please review this PR and tell me if any changes are required, I have added tests in the codebase for auto_host_rewrite

timflannagan commented:
@NishantSinghhhhh Looks like the formatting needs to be updated based on the changes to the api/applyconfiguration files, which are all auto-generated. You should be able to fix that issue by running the `generate-all` Makefile target locally.

yuval-k commented on test/kubernetes/e2e/features/auto_host_rewrite/input/backend.yaml:
```diff
@@ -0,0 +1,36 @@
+    spec:
+      containers:
+        - name: echo
+          image: hashicorp/http-echo
```
always pin images to a specific version, to prevent future CI breakage

NishantSinghhhhh replied:
Okay sir will remember this thing in future

lgadban commented:
It looks like there are still formatting changes in the generated files that shouldn't be there.

Can you share the specific command you used to run code generation?

yuval-k commented:
hi @NishantSinghhhhh you may have missed this but see this comment regarding the e2e test https://github.com/kgateway-dev/kgateway/pull/11262#discussion_r2103343856

NishantSinghhhhh commented:
> hi @NishantSinghhhhh you may have missed this but see this comment regarding the e2e test https://github.com/kgateway-dev/kgateway/pull/11262#discussion_r2103343856

Yes sir I have added these changes, currently I am trying to write e2e tests for autohostwrite, once it's done I will commit the changes

NishantSinghhhhh commented:
@yuval-k sir, the error that I am getting while running e2e tests are 

` kgateway git:(autoHostRewrite) ✗ go test -v -timeout 25m ./test/kubernetes/e2e/tests \                    
  -run '^TestKgateway$/^AutoHostRewrite$'                                      

=== RUN   TestKgateway
"Sun, 25 May 2025 22:51:03 IST: /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/helper/install.go:45"    found Helm index file at: "/home/nishant/LOCAL_DISK_D/Kgateway/kgateway/_test/index.yaml"
"Sun, 25 May 2025 22:51:03 IST: /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/helper/install.go:54"    version of ["kgateway-crds"] Helm chart is: "1.0.1-dev"
Release "kgateway-crds" does not exist. Installing it now.
NAME: kgateway-crds
LAST DEPLOYED: Sun May 25 22:51:04 2025
NAMESPACE: kgateway-test
STATUS: deployed
REVISION: 1
TEST SUITE: None
"Sun, 25 May 2025 22:51:04 IST: /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/helper/install.go:45"    found Helm index file at: "/home/nishant/LOCAL_DISK_D/Kgateway/kgateway/_test/index.yaml"
"Sun, 25 May 2025 22:51:04 IST: /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/helper/install.go:54"    version of ["kgateway"] Helm chart is: "1.0.1-dev"
Release "kgateway" does not exist. Installing it now.
NAME: kgateway
LAST DEPLOYED: Sun May 25 22:51:04 2025
NAMESPACE: kgateway-test
STATUS: deployed
REVISION: 1
TEST SUITE: None
2025-05-25T17:21:05.042905Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have status Running, but it was Pending
2025-05-25T17:21:07.051812Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
2025-05-25T17:21:09.060109Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
2025-05-25T17:21:11.066628Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
2025-05-25T17:21:13.075354Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
2025-05-25T17:21:15.081929Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
2025-05-25T17:21:17.089554Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
2025-05-25T17:21:19.097297Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
2025-05-25T17:21:21.105997Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
2025-05-25T17:21:23.112906Z     info    expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have condition ready, but it was not ready
    pods.go:71: 
        Timed out after 20.000s.
        Failed to match pod in namespace kgateway-test, with conditions {{ } app.kubernetes.io/name=kgateway  false false   <nil> 0  <nil>}
        The function passed to Eventually failed at /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/assertions/pods.go:66 with:
        Expected pod kgateway-test/kgateway-7b8d99b845-2sz8d to have status 'Running', but it was not found
Test failed. Dumping state from default, kgateway-test, kube-node-lease, kube-public, kube-system, local-path-storage...
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
Finished dumping kubernetes state
no kgateway=kgateway pods found in namespace default
no kgateway=kgateway pods found in namespace default
found controller pods: kgateway-7b8d99b845-2sz8d
error running krt snapshot command: 
error running xds snapshot command: 
finished dumping controller state
no kgateway=kgateway pods found in namespace kube-node-lease
no kgateway=kgateway pods found in namespace kube-node-lease
no kgateway=kgateway pods found in namespace kube-public
no kgateway=kgateway pods found in namespace kube-public
no kgateway=kgateway pods found in namespace kube-system
no kgateway=kgateway pods found in namespace kube-system
no kgateway=kgateway pods found in namespace local-path-storage
no kgateway=kgateway pods found in namespace local-path-storage
no kgateway=kube-gateway pods found in namespace default
no proxies found in namespace default
no kgateway=kube-gateway pods found in namespace kgateway-test
no proxies found in namespace kgateway-test
no kgateway=kube-gateway pods found in namespace kube-node-lease
no proxies found in namespace kube-node-lease
no kgateway=kube-gateway pods found in namespace kube-public
no proxies found in namespace kube-public
no kgateway=kube-gateway pods found in namespace kube-system
no proxies found in namespace kube-system
no kgateway=kube-gateway pods found in namespace local-path-storage
no proxies found in namespace local-path-storage
Test failed. Logs and cluster state are available in /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/_test/bug_report/kind/kgateway-test
release "kgateway" uninstalled
release "kgateway-crds" uninstalled
--- FAIL: TestKgateway (29.39s)
FAIL
FAIL    github.com/kgateway-dev/kgateway/v2/test/kubernetes/e2e/tests   29.418s
FAIL
➜  kgateway git:(autoHostRewrite) ✗ `

can you please tell me how can I solve these errors

NishantSinghhhhh commented:
@yuval-k sir, the error that I am getting while running e2e tests are 

` kgateway git:(autoHostRewrite) ✗ go test -v -timeout 25m ./test/kubernetes/e2e/tests \                    
  -run '^TestKgateway$/^AutoHostRewrite$'                                      
 kgateway git:(autoHostRewrite) ✗ go test -v -timeout 25m ./test/kubernetes/e2e/tests \
  -run '^TestKgateway$/^AutoHostRewrite$'

=== RUN   TestKgateway
"Sun, 25 May 2025 23:06:01 IST: /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/helper/install.go:45"    found Helm index file at: "/home/nishant/LOCAL_DISK_D/Kgateway/kgateway/_test/index.yaml"
"Sun, 25 May 2025 23:06:01 IST: /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/helper/install.go:54"    version of ["kgateway-crds"] Helm chart is: "1.0.1-dev"
Release "kgateway-crds" does not exist. Installing it now.
NAME: kgateway-crds
LAST DEPLOYED: Sun May 25 23:06:01 2025
NAMESPACE: kgateway-test
STATUS: deployed
REVISION: 1
TEST SUITE: None
"Sun, 25 May 2025 23:06:01 IST: /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/helper/install.go:45"    found Helm index file at: "/home/nishant/LOCAL_DISK_D/Kgateway/kgateway/_test/index.yaml"
"Sun, 25 May 2025 23:06:01 IST: /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/helper/install.go:54"    version of ["kgateway"] Helm chart is: "1.0.1-dev"
Release "kgateway" does not exist. Installing it now.
NAME: kgateway
LAST DEPLOYED: Sun May 25 23:06:01 2025
NAMESPACE: kgateway-test
STATUS: deployed
REVISION: 1
TEST SUITE: None
    pods.go:71: 
        Timed out after 20.000s.
        Failed to match pod in namespace kgateway-test, with conditions {{ } app.kubernetes.io/name=kgateway  false false   <nil> 0  <nil>}
        The function passed to Eventually failed at /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/test/kubernetes/testutils/assertions/pods.go:64 with:
        No pods found
        Expected
            <[]v1.Pod | len:0, cap:0>: nil
        not to be empty
Test failed. Dumping state from default, kgateway-test, kube-node-lease, kube-public, kube-system, local-path-storage...
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in default namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kgateway-test namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-node-lease namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-public namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in kube-system namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
  No resources found in local-path-storage namespace.
Finished dumping kubernetes state
no kgateway=kgateway pods found in namespace default
no kgateway=kgateway pods found in namespace default
found controller pods: kgateway-7b8d99b845-hcxnh
error running krt snapshot command: 
error running xds snapshot command: 
finished dumping controller state
no kgateway=kgateway pods found in namespace kube-node-lease
no kgateway=kgateway pods found in namespace kube-node-lease
no kgateway=kgateway pods found in namespace kube-public
no kgateway=kgateway pods found in namespace kube-public
no kgateway=kgateway pods found in namespace kube-system
no kgateway=kgateway pods found in namespace kube-system
no kgateway=kgateway pods found in namespace local-path-storage
no kgateway=kgateway pods found in namespace local-path-storage
no kgateway=kube-gateway pods found in namespace default
no proxies found in namespace default
no kgateway=kube-gateway pods found in namespace kgateway-test
no proxies found in namespace kgateway-test
no kgateway=kube-gateway pods found in namespace kube-node-lease
no proxies found in namespace kube-node-lease
no kgateway=kube-gateway pods found in namespace kube-public
no proxies found in namespace kube-public
no kgateway=kube-gateway pods found in namespace kube-system
no proxies found in namespace kube-system
no kgateway=kube-gateway pods found in namespace local-path-storage
no proxies found in namespace local-path-storage
Test failed. Logs and cluster state are available in /home/nishant/LOCAL_DISK_D/Kgateway/kgateway/_test/bug_report/kind/kgateway-test
release "kgateway" uninstalled
release "kgateway-crds" uninstalled
--- FAIL: TestKgateway (26.03s)
FAIL
FAIL    github.com/kgateway-dev/kgateway/v2/test/kubernetes/e2e/tests   26.061s
FAIL

can you please tell me how can I solve these errors

NishantSinghhhhh commented:
> > Failed to match pod in namespace kgateway-test, with conditions {{ } app.kubernetes.io/name=kgateway false false 0 }
> 
> @NishantSinghhhhh Hmm, that seems like a bug. Try changing the default namespace before running the e2e suite. https://github.com/ahmetb/kubectx is what I'd recommend for doing this.

[Screencast from 2025-01-21 12-27-42.webm](https://github.com/user-attachments/assets/f7b992ec-9229-430a-a7d8-1abbb94a5baf)

Sir I am getting the same error again

timflannagan commented:
@NishantSinghhhhh let's also talk in the cncf slack channel if you run into more issues that block progress on this

NishantSinghhhhh commented:
@timflannagan sir https://cloud-native.slack.com/archives/C080D3PJMS4/p1748556644213119, I have asked my doubts in this slack thread, can you please help me out with this

NishantSinghhhhh commented:
@timflannagan sir https://cloud-native.slack.com/archives/C080D3PJMS4/p1748556644213119, I have asked my doubts in this slack thread, can you  help me out with this

npolshakova commented:
Looks like the tests are passing, but there are some static code / lint fixes that need to be made: https://github.com/kgateway-dev/kgateway/actions/runs/15367018356/job/43260193267?pr=11262

timflannagan commented:
@NishantSinghhhhh I just pulled down the branch locally and ran the following commands:

```bash
$ make generate-all ; git status
...
On branch autoHostRewrite
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   api/applyconfiguration/api/v1alpha1/awsauth.go
        modified:   api/applyconfiguration/api/v1alpha1/backend.go
        modified:   api/applyconfiguration/api/v1alpha1/directresponse.go
        modified:   api/applyconfiguration/api/v1alpha1/extauthpolicy.go
        modified:   api/applyconfiguration/api/v1alpha1/gatewayextension.go
        modified:   api/applyconfiguration/api/v1alpha1/gatewayparameters.go
        modified:   api/applyconfiguration/api/v1alpha1/httplistenerpolicy.go
        modified:   api/applyconfiguration/api/v1alpha1/service.go
        modified:   api/applyconfiguration/api/v1alpha1/singleauthtoken.go
        modified:   api/applyconfiguration/api/v1alpha1/trafficpolicy.go
        modified:   api/applyconfiguration/internal/internal.go
        modified:   api/applyconfiguration/utils.go
        modified:   go.sum
        ...
```

You can also run the `make verify` target, which re-generates everything and then runs git diff against those changes. This is the same check that CI is running here.

Additionally, the `make analyze` target will point any additional issues that would fail static validation. I ran that locally on this branch and saw a couple of linting violations pop up too:

```bash
$ make analyze
...
internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:14:2: ST1019: package "github.com/envoyproxy/go-control-plane/envoy/config/route/v3" is being imported more than once (staticcheck)
        route "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
        ^
internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:15:2: ST1019(related information): other import of "github.com/envoyproxy/go-control-plane/envoy/config/route/v3" (staticcheck)
        routev3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
        ^
internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:143:2: field errors is unused (unused)
        errors                     []error
        ^
test/kubernetes/e2e/features/auto_host_rewrite/suite.go:23:7: const namespace is unused (unused)
const namespace = "auto-host-rewrite"
      ^
4 issues:
* staticcheck: 2
* unused: 2
```

Let me know if you run into any issues with those targets.

NishantSinghhhhh commented:
@yuval-k Sir, the error in the test workflow is coming from a file I haven't modified. Could you please guide me on what changes I should make to resolve this and pass the test workflow?

timflannagan commented:
@NishantSinghhhhh No need to worry about it -- known flake we're investigating further.

puertomontt commented on api/v1alpha1/traffic_policy_types.go:
```diff
@@ -72,6 +72,11 @@ type TrafficPolicySpec struct {
+
+	// AutoHostRewrite rewrites the Host header to the DNS name of the selected upstream.
+	// Only honoured for HTTPRoute targets.
+	// +kubebuilder:validation:Enum=true;false
```
this validation is unnecessary since k8s will already reject values that aren't true/false.
```suggestion
```

lgadban commented on internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:
```diff
@@ -196,10 +199,15 @@ func (d *TrafficPolicy) Equals(in any) bool {
 		return false
 	}
 
-	if !d.spec.cors.Equals(d2.spec.cors) {
```
is this removing the cors Equals() check?
we need to add this back in if so

NishantSinghhhhh replied:
Have added this in the recent commit

lgadban commented on internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:
```diff
@@ -1085,6 +1101,11 @@ func (b *TrafficPolicyBuilder) Translate(
 		errors = append(errors, err)
 	}
 
+	// 🆕 Handle autoHostRewrite
```
unnecessary comment (and likely we won't want emoji comments either)

NishantSinghhhhh commented:
@yuval-k yes sir , The file was changed completely so finding a way to add autohostrewrite in the new code

NishantSinghhhhh commented:
@yuval-k sir, I have added the logic for autohostrewrite in traffic_policy_plugin.go, Can you have a look at the changes

1. IR struct update : Added autoHostRewrite *wrapperspb.BoolValue to trafficPolicySpecIr.

2. Equality logic : Extended TrafficPolicy.Equals to compare spec.autoHostRewrite via proto.Equal.

3. Builder (Translate) : When the CR’s spec.autoHostRewrite is true, sets outSpec.autoHostRewrite = wrapperspb.Bool(true).

4. Per-route injection (ApplyForRoute) : If the flag is true, assigns
RouteAction.HostRewriteSpecifier = RouteAction_AutoHostRewrite{AutoHostRewrite: …}.

5. VirtualHost-wide loop (ApplyVhostPlugin) : When the flag is true, iterates over out.Routes and stamps the same AutoHostRewrite specifier onto every RouteAction.

6. Merge logic (MergeTrafficPolicies) : Added an IsMergeable block that copies autoHostRewrite and records mergeOrigins["autoHostRewrite"].

NishantSinghhhhh commented:
> Hi @NishantSinghhhhh thanks for sticking with this effort, it is getting close!
> 
> One thing I noticed is that the new autoHostRewrite logic is absent from the [`ApplyRouteConfigPlugin` plugin method](https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go#L241), IIRC this might be needed for the policy to work correctly when the TrafficPolicy is attached to a Gateway only.
> 
> A good way to check this is with the setup test suite which allows for quickly testing the output of the translation pass for different scenarios, without the overhead of an e2e test.
> 
> Can you add a setup test, similar to this [local rate limit test](https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/setup/testdata/standard/local-rate-limit-route-gw.yaml) which will allow us to inspect the xds output and make sure it is working as intended?

yes sir I will add a test similar to the rate limit for autoHostRewrite, and will also add the logic in [ApplyRouteConfigPlugin plugin method](https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go#L241), IIRC

yuval-k commented:
hi @NishantSinghhhhh you don't need to push merge commits unless there's a conflict or you think there is a CI flake (usually in the end- to-end  checks). 
We have a merge queue set-up so, so running CI against the merged version will be done automatically. 
the downside of pushing the merge commits is that it re-sets the CI status.

lgadban reviewed (commented):
With the new test, you will need to actually run the suite to get a corresponding `*-out.yaml` file that needs to be added and checked in to the repo, like [this one](https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/setup/testdata/standard/cors-trafficpolicy-gw-out.yaml).

You can follow these directions to run the tests: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/setup/README_test.md#how-to-run

lgadban commented:
@NishantSinghhhhh you also have failures that need to resolved by running codegen like before, `make generated-code`

lgadban commented on internal/kgateway/setup/testdata/standard/auto-host-rewrite-out.yaml:
```diff
@@ -0,0 +1,54 @@
+          typedConfig:
+            '@type': type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+
+routes:
+- name: listener~8080
+  ignorePortInHostMatching: true
```
this doesn't seem correct. something is not right with this test as no routes are being reported

NishantSinghhhhh commented:
@yuval-k, @lgadban sir I have added changes so that tests can be reported, can you tell me if the changes done are correct or not

NishantSinghhhhh commented:
@yuval-k, @lgadban sir I have added changes so that tests can be reported, can you tell me if the changes done are correct or not

NishantSinghhhhh commented on internal/kgateway/setup/testdata/standard/auto-host-rewrite-out.yaml:
```diff
@@ -0,0 +1,64 @@
+  metadata: {}
+  name: kube_default_kubernetes_443
+  type: EDS
+
```
Sir 
You're right about the newlines. This was generated by running the setup tests. Should I investigate the YAML generation code or use a different test command?

lgadban commented:
Hi @NishantSinghhhhh -- I ended up pushing a few changes in the interest of us getting this PR merged.

Thanks for the contribution and getting this one started!

timflannagan commented on internal/kgateway/extensions2/plugins/trafficpolicy/auto_host_rewrite_test.go:
```diff
@@ -0,0 +1,56 @@
+	ctx := context.Background()
+	plugin := &trafficPolicyPluginGwPass{}
+
+	t.Run("autoHostRewrite true → RouteAction flag set", func(t *testing.T) {
```
nit: rip out the `→` model generated flag?

timflannagan commented on internal/kgateway/extensions2/plugins/trafficpolicy/builder.go:
```diff
@@ -104,6 +105,10 @@ func (b *TrafficPolicyBuilder) Translate(
 		errors = append(errors, err)
 	}
 
+	if policyCR.Spec.AutoHostRewrite != nil {
+		outSpec.autoHostRewrite = wrapperspb.Bool(*policyCR.Spec.AutoHostRewrite)
+	}
```
nit: Creating a dedicated IR type would be beneficial for the TP validate pattern we were discussing last week.

timflannagan commented on internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go:
```diff
@@ -361,6 +374,15 @@ func (p *trafficPolicyPluginGwPass) ApplyForRoute(ctx context.Context, pCtx *ir.
 		}
 	}
+
+	if policy.spec.autoHostRewrite != nil && policy.spec.autoHostRewrite.GetValue() {
+		if ra := outputRoute.GetRoute(); ra != nil {
+			ra.HostRewriteSpecifier = &routev3.RouteAction_AutoHostRewrite{
+				AutoHostRewrite: policy.spec.autoHostRewrite,
+			}
+		}
+	}
```
Do we need to handle the scenario where the output route already has an action configured?

timflannagan commented on test/kubernetes/e2e/defaults/defaults.go:
```diff
@@ -92,7 +92,7 @@ Commercial support is available at
 	HttpbinPod = &corev1.Pod{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      "httpbin",
+			Name:      "httpbin", // FIXME: this is incorrect, pod name is ~generated
```
super nit: s/FIXME/TODO. IMO, FIXME should be reserved for things that should block a PR from merging.

timflannagan commented on test/kubernetes/e2e/features/auto_host_rewrite/suite.go:
```diff
@@ -0,0 +1,106 @@
+	}
+}
+
+/* ───────────────────────── Set-up / Tear-down ───────────────────────── */
```
nit: rip this marker out? relevant for line 69 as well

timflannagan commented on test/kubernetes/e2e/features/auto_host_rewrite/suite.go:
```diff
@@ -0,0 +1,106 @@
+		LabelSelector: defaults.HttpbinLabelSelector,
+	})
+	s.ti.Assertions.EventuallyPodsRunning(s.ctx, proxyObjectMeta.GetNamespace(), metav1.ListOptions{
+		LabelSelector: defaults.WellKnownAppLabel + "=" + proxyObjectMeta.GetName(),
```
TIL this defaults.WellKnownAppLabel exists.

timflannagan commented on test/kubernetes/e2e/features/auto_host_rewrite/types.go:
```diff
@@ -0,0 +1,39 @@
+var (
+	autoHostRewriteManifest = filepath.Join(fsutils.MustGetThisDir(), "testdata", "auto_host_rewrite.yaml")
+
+	/* objects from gateway manifest (gw.yaml) */
```
nit: rip this comment and the one line on line 26 out.

timflannagan commented on internal/kgateway/setup/testdata/standard/auto-host-plus-host-rewrite-out.yaml:
Move this and the other test case to the gateway translator suite?

timflannagan commented on api/v1alpha1/traffic_policy_types.go:
```diff
@@ -81,6 +82,13 @@ type TrafficPolicySpec struct {
 	// +optional
 	Csrf *CSRFPolicy `json:"csrf,omitempty"`
 
+	// AutoHostRewrite rewrites the Host header to the DNS name of the selected upstream.
+	// NOTE: This field is only honoured for HTTPRoute targets.
+	// NOTE: If `autoHostRewrite` is set on a route that also has a [URLRewrite filter](https://gateway-api.sigs.k8s.io/reference/spec/#httpurlrewritefilter)
+	// configured to override the `hostname`, the `hostname` value will be used and `autoHostRewrite` will be ignored.
+	// +optional
+	AutoHostRewrite *bool `json:"autoHostRewrite,omitempty"`
```
I think we could improve the UX for this scenario, but it's an existing problem so nothing that needs to be tackled here.

timflannagan reviewed (commented):
Quick passthrough @lgadban - mostly nits, but can approve. The only thing that could be good to tackle is moving those test cases to gateway translator suite if possible.

timflannagan reviewed (approved):
Let's tackle any follow-ups in a separate PR. @NishantSinghhhhh thanks for the initial contribution and @lgadban for getting this over the line.
